src/common/io/socket/common.c

```diff
@@ -126,3 +128,86 @@ sckOptionSet(int fd)
 
     FUNCTION_TEST_RETURN_VOID();
 }
+
+/***********************************************************************************************************************************
+Use poll() to determine when data is ready to read/write on a socket. Retry after EINTR with whatever time is left on the timer.
+***********************************************************************************************************************************/
+// Helper to determine when poll() should be retried
+static bool
+sckReadyRetry(int pollResult, int errNo, bool first, Wait *wait)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(INT, pollResult);
+        FUNCTION_TEST_PARAM(INT, errNo);
+        FUNCTION_TEST_PARAM(BOOL, first);
+        FUNCTION_TEST_PARAM(WAIT, wait);
+    FUNCTION_TEST_END();
+
+    // No retry by default
+    bool result = false;
+
+    // Process errors
+    if (pollResult == -1)
+    {
+        // Don't error on an interrupt. If the interrupt lasts long enough there may be a timeout, though.
+        if (errNo != EINTR)
+            THROW_SYS_ERROR_CODE(errNo, KernelError, "unable to poll socket");
+
+        // Retry if first iteration or time remaining
+        result = first || waitMore(wait);
+    }
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+bool
+sckReady(int fd, bool read, bool write, TimeMSec timeout)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(INT, fd);
+        FUNCTION_LOG_PARAM(BOOL, read);
+        FUNCTION_LOG_PARAM(BOOL, write);
+        FUNCTION_LOG_PARAM(TIME_MSEC, timeout);
+    FUNCTION_LOG_END();
+
+    ASSERT(fd != -1);
+    ASSERT(read || write);
+    ASSERT(timeout < INT_MAX);
+
+    // Poll settings
+    struct pollfd inputFd = {.fd = fd, .events = POLLERR};
```

Comment:
I don't understand .event=POLLERR because the documentation states:
POLLERR
    An error has occurred on the device or stream. This flag is only valid in the **_revents_** bitmask; it shall be **ignored in the _events_ member**

So why is it set here?

Reply:
I left this in because PG does it even though as you say the Linux man page says it is useless. I checked BSD and it behaves the same way, but it could be this is defensive programming for some older non-Posix-compliant version of poll().

My vote is to leave it in and add a comment as above.

Reply:
Well, Stephen noted the same thing, so I'll kill it despite what PG does. We really aren't interested in non-Posix versions of poll().